graph/src/data_source/offchain.rs

```diff
@@ -32,35 +33,34 @@ pub struct DataSource {
     pub mapping: Mapping,
     pub context: Arc<Option<DataSourceContext>>,
     pub creation_block: Option<BlockNumber>,
-    pub done_at: Mutex<Option<i32>>,
+    pub done_at: Arc<AtomicI32>,
     pub causality_region: CausalityRegion,
 }
 
-impl Clone for DataSource {
-    fn clone(&self) -> Self {
-        Self {
-            kind: self.kind.clone(),
-            name: self.name.clone(),
-            manifest_idx: self.manifest_idx.clone(),
-            source: self.source.clone(),
-            mapping: self.mapping.clone(),
-            context: self.context.clone(),
-            creation_block: self.creation_block.clone(),
-            done_at: Mutex::new(*self.done_at.lock().unwrap()),
-            causality_region: self.causality_region.clone(),
-        }
-    }
-}
-
 impl DataSource {
     // mark this data source as processed.
     pub fn mark_processed_at(&self, block_no: i32) {
-        *self.done_at.lock().unwrap() = Some(block_no);
+        self.done_at
+            .store(block_no, std::sync::atomic::Ordering::SeqCst);
     }
 
     // returns `true` if the data source is processed.
     pub fn is_processed(&self) -> bool {
-        self.done_at.lock().unwrap().is_some()
+        self.done_at.load(std::sync::atomic::Ordering::SeqCst) != NOT_DONE_VALUE
+    }
+
+    pub fn done_at(&self) -> Option<i32> {
+        match self.done_at.load(std::sync::atomic::Ordering::SeqCst) {
+            NOT_DONE_VALUE => None,
+            n => Some(n),
+        }
+    }
+
+    pub fn set_at(&self, block: Option<i32>) {
```

Comment:
```suggestion
    pub fn set_done_at(&self, block: Option<i32>) {
```

Reply:
I'm very confused as to how the done got lost :thinking: